facebookresearch/ParlAI
# Add support for images to WebSockets chat service

**Patch description**
This PR adds support for images to the WebSockets chat service. Given this, there are changes to the message structure while attempting to follow the chat service message structure defined in the `chat_service` README as close as possible. Additionally, this adds support for a `module_initialize` function in the worlds file.

The `text` field in the messages is always a string of text, which makes it easy to compare the text in a message to a string (or to support a single message with text + image in the future).
The `attachment` field is for images, where the following fields are expected:
- type (the only supported type right now is `image`)
- text  (For images, this is the base 64 encoded image)
- mime_type (This is the Mime type for the images, this is necessary for an HTML UI to render an image correctly)

WebSocket messages are now stringified JSON objects that conform to the message structure above.

**Testing steps**
See #2132 for running a websocket server for testing. Messages must be sent as stringifed JSON now using the structure above. To send images, provide the `attachment` object with a base64 encoded image for the text.

Example text message:
'{"text": "Example text message"}'
Example image message:
'{"attachment": {"type": "image", "text": "<IMAGE BASE64>", "mime_type": "image/jpg"}}'
